nomos-services/cryptarchia-consensus/src/lib.rs

```diff
@@ -872,6 +854,206 @@ where
             blobs = blobs
         );
     }
+
+    /// Retrieves the blocks in the range from `from` to `to` from the storage.
+    /// Both `from` and `to` are included in the range.
+    /// This is implemented here, and not as a method of `StorageAdapter`, to simplify the panic
+    /// and error message handling.
+    ///
+    /// # Panics
+    ///
+    /// Panics if any of the blocks in the range are not found in the storage.
+    ///
+    /// # Parameters
+    ///
+    /// * `from` - The header id of the first block in the range. Must be a valid header.
+    /// * `to` - The header id of the last block in the range. Must be a valid header.
+    ///
+    /// # Returns
+    ///
+    /// A vector of blocks in the range from `from` to `to`.
+    /// If no blocks are found, returns an empty vector.
+    /// If any of the HeaderId are invalid, returns an error with the first invalid header id.
+    async fn get_blocks_in_range(
+        from: HeaderId,
+        to: HeaderId,
+        storage_adapter: &StorageAdapter<Storage, TxS::Tx, BS::BlobId>,
+    ) -> Vec<Block<ClPool::Item, DaPool::Item>> {
+        // Due to the blocks traversal order, this yields `to..from` order
+        let blocks = futures::stream::unfold(to, |header_id| async move {
+            if header_id == from {
+                None
+            } else {
+                let block = storage_adapter
+                    .get_block(&header_id)
+                    .await
+                    .unwrap_or_else(|| {
+                        panic!("Could not retrieve block {to} from storage during recovery")
+                    });
+                let parent_header_id = block.header().parent();
+                Some((block, parent_header_id))
+            }
+        });
+
+        // To avoid confusion, the order is reversed so it fits the natural `from..to` order
+        blocks.collect::<Vec<_>>().await.into_iter().rev().collect()
```

Comment:
I think it would be better to return the stream here instead of the collected vec. Then collect it externally when used. Is more generic case imo.

Reply:
The reason for this is to force it to natural order, because afaik, you can't reverse an stream.
Due to `unfold`'s nature, items are in `to` to `from` order, which could be confusing to callers.

Reply:
There is an [`iter.rev()`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.rev) which might be what you are looking for? This works on iterators, not sure if streams are something else?

Reply:
That's probably what I'm already using in the `into_iter().rev()` section. It makes sense to me that you can't chain a `Stream` with a `rev` without explicitly storing it first: You need to store the whole stream in memory (e.g.: vector) before reversing it, otherwise how do you know what stream item is the first one?